File: src/model/reference/airmass_phase.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const PhaseSupportKind = enum {
    none,
    analytic_hg,
    mie_table,
};

pub const AirmassFactorPoint = struct {
    solar_zenith_deg: f64,
    view_zenith_deg: f64,
    relative_azimuth_deg: f64,
    airmass_factor: f64,
};

pub const MiePhasePoint = struct {
    wavelength_nm: f64,
    extinction_scale: f64,
    single_scatter_albedo: f64,
    phase_coefficients: [4]f64,
};

pub const MiePhaseTable = struct {
    points: []MiePhasePoint,

    pub fn deinit(self: *MiePhaseTable, allocator: Allocator) void {
        allocator.free(self.points);
        self.* = undefined;
    }

    pub fn interpolate(self: MiePhaseTable, wavelength_nm: f64) MiePhasePoint {
        if (self.points.len == 0) {
            return .{
                .wavelength_nm = wavelength_nm,
                .extinction_scale = 1.0,
                .single_scatter_albedo = 1.0,
                .phase_coefficients = .{ 1.0, 0.0, 0.0, 0.0 },
            };
        }
        if (wavelength_nm <= self.points[0].wavelength_nm) return self.points[0];

        for (self.points[0 .. self.points.len - 1], self.points[1..]) |left, right| {
            if (wavelength_nm <= right.wavelength_nm) {
                const span = right.wavelength_nm - left.wavelength_nm;
                if (span == 0.0) return right;
                const weight = (wavelength_nm - left.wavelength_nm) / span;

                var phase_coefficients: [4]f64 = undefined;
                for (&phase_coefficients, 0..) |*slot, index| {
                    slot.* = left.phase_coefficients[index] +
                        weight * (right.phase_coefficients[index] - left.phase_coefficients[index]);
                }
                phase_coefficients[0] = 1.0;
                return .{
                    .wavelength_nm = wavelength_nm,
                    .extinction_scale = left.extinction_scale + weight * (right.extinction_scale - left.extinction_scale),
                    .single_scatter_albedo = left.single_scatter_albedo + weight * (right.single_scatter_albedo - left.single_scatter_albedo),
                    .phase_coefficients = phase_coefficients,
                };
            }
        }
        return self.points[self.points.len - 1];
    }
};

pub const AirmassFactorLut = struct {
    points: []AirmassFactorPoint,

    pub fn deinit(self: *AirmassFactorLut, allocator: Allocator) void {
        allocator.free(self.points);
        self.* = undefined;
    }

    pub fn nearest(self: AirmassFactorLut, solar_zenith_deg: f64, view_zenith_deg: f64, relative_azimuth_deg: f64) f64 {
        if (self.points.len == 0) return 1.0;

        var best_distance = std.math.inf(f64);
        var best_value = self.points[0].airmass_factor;
        for (self.points) |point| {
            const delta_sza = point.solar_zenith_deg - solar_zenith_deg;
            const delta_vza = point.view_zenith_deg - view_zenith_deg;
            const delta_raa = point.relative_azimuth_deg - relative_azimuth_deg;
            const distance = delta_sza * delta_sza + delta_vza * delta_vza + delta_raa * delta_raa;
            if (distance < best_distance) {
                best_distance = distance;
                best_value = point.airmass_factor;
            }
        }
        return best_value;
    }

    pub fn providesSupportOnly(_: AirmassFactorLut) bool {
        return true;
    }
};

pub fn spectralProfileFromOpticalDepth(
    allocator: Allocator,
    wavelengths_nm: []const f64,
    mean_airmass_factor: f64,
    optical_depth_proxy: []const f64,
) ![]f64 {
    if (wavelengths_nm.len != optical_depth_proxy.len) return error.ShapeMismatch;

    const profile = try allocator.alloc(f64, wavelengths_nm.len);
    errdefer allocator.free(profile);
    if (wavelengths_nm.len == 0) return profile;

    var proxy_sum: f64 = 0.0;
    for (optical_depth_proxy) |value| proxy_sum += @max(value, 0.0);
    const proxy_mean = proxy_sum / @max(@as(f64, @floatFromInt(optical_depth_proxy.len)), 1.0e-9);
    const safe_mean_airmass = if (std.math.isFinite(mean_airmass_factor) and mean_airmass_factor > 0.0)
        mean_airmass_factor
    else
        1.0;
    const midpoint_nm = 0.5 * (wavelengths_nm[0] + wavelengths_nm[wavelengths_nm.len - 1]);
    const half_span_nm = @max(0.5 * (wavelengths_nm[wavelengths_nm.len - 1] - wavelengths_nm[0]), 1.0e-9);

    for (profile, wavelengths_nm, optical_depth_proxy) |*slot, wavelength_nm, proxy| {
        const normalized_proxy = if (proxy_mean > 0.0) @max(proxy, 0.0) / proxy_mean else 1.0;
        const coordinate = (wavelength_nm - midpoint_nm) / half_span_nm;
        // DECISION:
        //   Keep a weak wavelength tilt so the profile is not numerically flat before the mean is restored.
        const geometric_tilt = 1.0 + 0.05 * coordinate;
        slot.* = safe_mean_airmass * normalized_proxy * geometric_tilt;
    }

    var current_mean: f64 = 0.0;
    for (profile) |value| current_mean += value;
    current_mean /= @max(@as(f64, @floatFromInt(profile.len)), 1.0e-9);
    const renormalization = safe_mean_airmass / @max(current_mean, 1.0e-9);
    for (profile) |*value| value.* *= renormalization;
    return profile;
}

test "spectral amf profile preserves the requested mean factor" {
    const wavelengths = [_]f64{ 759.0, 760.0, 761.0, 762.0 };
    const proxy = [_]f64{ 0.5, 1.0, 1.5, 1.0 };
    const profile = try spectralProfileFromOpticalDepth(std.testing.allocator, &wavelengths, 2.0, &proxy);
    defer std.testing.allocator.free(profile);

    var mean: f64 = 0.0;
    for (profile) |value| mean += value;
    mean /= @as(f64, @floatFromInt(profile.len));
    try std.testing.expectApproxEqAbs(@as(f64, 2.0), mean, 1.0e-9);
    try std.testing.expect(profile[2] > profile[0]);
}

File: rust_src/input/reference/airmass_phase.rs
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub enum PhaseSupportKind {
    #[default]
    None,
    AnalyticHg,
    MieTable,
}

#[derive(Debug, Default, Clone, Copy, PartialEq)]
pub struct AirmassFactorPoint {
    pub solar_zenith_deg: f64,
    pub view_zenith_deg: f64,
    pub relative_azimuth_deg: f64,
    pub airmass_factor: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct MiePhasePoint {
    pub wavelength_nm: f64,
    pub extinction_scale: f64,
    pub single_scatter_albedo: f64,
    pub phase_coefficients: [f64; 4],
}

impl Default for MiePhasePoint {
    fn default() -> Self {
        Self {
            wavelength_nm: 0.0,
            extinction_scale: 1.0,
            single_scatter_albedo: 1.0,
            phase_coefficients: [1.0, 0.0, 0.0, 0.0],
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct MiePhaseTable {
    pub points: Vec<MiePhasePoint>,
}

impl MiePhaseTable {
    pub fn interpolate(&self, wavelength_nm: f64) -> MiePhasePoint {
        if self.points.is_empty() {
            return MiePhasePoint {
                wavelength_nm,
                ..MiePhasePoint::default()
            };
        }
        if wavelength_nm <= self.points[0].wavelength_nm {
            return self.points[0];
        }

        for pair in self.points.windows(2) {
            let left = pair[0];
            let right = pair[1];
            if wavelength_nm > right.wavelength_nm {
                continue;
            }
            let span = right.wavelength_nm - left.wavelength_nm;
            if span == 0.0 {
                return right;
            }
            let weight = (wavelength_nm - left.wavelength_nm) / span;
            let mut phase_coefficients = [0.0; 4];
            for (index, slot) in phase_coefficients.iter_mut().enumerate() {
                *slot = left.phase_coefficients[index]
                    + weight * (right.phase_coefficients[index] - left.phase_coefficients[index]);
            }
            phase_coefficients[0] = 1.0;
            return MiePhasePoint {
                wavelength_nm,
                extinction_scale: left.extinction_scale
                    + weight * (right.extinction_scale - left.extinction_scale),
                single_scatter_albedo: left.single_scatter_albedo
                    + weight * (right.single_scatter_albedo - left.single_scatter_albedo),
                phase_coefficients,
            };
        }
        *self.points.last().expect("non-empty Mie phase table")
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct AirmassFactorLut {
    pub points: Vec<AirmassFactorPoint>,
}

impl AirmassFactorLut {
    pub fn nearest(
        &self,
        solar_zenith_deg: f64,
        view_zenith_deg: f64,
        relative_azimuth_deg: f64,
    ) -> f64 {
        if self.points.is_empty() {
            return 1.0;
        }

        let mut best_distance = f64::INFINITY;
        let mut best_value = self.points[0].airmass_factor;
        for point in &self.points {
            let delta_sza = point.solar_zenith_deg - solar_zenith_deg;
            let delta_vza = point.view_zenith_deg - view_zenith_deg;
            let delta_raa = point.relative_azimuth_deg - relative_azimuth_deg;
            let distance = delta_sza * delta_sza + delta_vza * delta_vza + delta_raa * delta_raa;
            if distance < best_distance {
                best_distance = distance;
                best_value = point.airmass_factor;
            }
        }
        best_value
    }

    pub fn provides_support_only(&self) -> bool {
        true
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    ShapeMismatch,
}

pub fn spectral_profile_from_optical_depth(
    wavelengths_nm: &[f64],
    mean_airmass_factor: f64,
    optical_depth_proxy: &[f64],
) -> Result<Vec<f64>, Error> {
    if wavelengths_nm.len() != optical_depth_proxy.len() {
        return Err(Error::ShapeMismatch);
    }
    if wavelengths_nm.is_empty() {
        return Ok(Vec::new());
    }

    let proxy_sum: f64 = optical_depth_proxy.iter().map(|value| value.max(0.0)).sum();
    let proxy_mean = proxy_sum / (optical_depth_proxy.len() as f64).max(1.0e-9);
    let safe_mean_airmass = if mean_airmass_factor.is_finite() && mean_airmass_factor > 0.0 {
        mean_airmass_factor
    } else {
        1.0
    };
    let midpoint_nm = 0.5 * (wavelengths_nm[0] + wavelengths_nm[wavelengths_nm.len() - 1]);
    let half_span_nm =
        (0.5 * (wavelengths_nm[wavelengths_nm.len() - 1] - wavelengths_nm[0])).max(1.0e-9);

    let mut profile = Vec::with_capacity(wavelengths_nm.len());
    for (wavelength_nm, proxy) in wavelengths_nm.iter().zip(optical_depth_proxy.iter()) {
        let normalized_proxy = if proxy_mean > 0.0 {
            proxy.max(0.0) / proxy_mean
        } else {
            1.0
        };
        let coordinate = (wavelength_nm - midpoint_nm) / half_span_nm;
        // The small tilt keeps generated support profiles from becoming flat,
        // then the final normalization restores the requested mean.
        let geometric_tilt = 1.0 + 0.05 * coordinate;
        profile.push(safe_mean_airmass * normalized_proxy * geometric_tilt);
    }

    let current_mean = profile.iter().sum::<f64>() / (profile.len() as f64).max(1.0e-9);
    let renormalization = safe_mean_airmass / current_mean.max(1.0e-9);
    for value in &mut profile {
        *value *= renormalization;
    }
    Ok(profile)
}
